src/api/docker-files/Dockerfile

```diff
@@ -15,15 +15,13 @@ RUN usermod -u $CONTAINER_USERID frontend
 # docker would use it's cache for this and the following stages.
 ADD Gemfile /obs/src/api/Gemfile
 ADD Gemfile.lock /obs/src/api/Gemfile.lock
+ADD .bundle /obs/src/api/
 RUN chown -R frontend /obs/src/api
 
 # Now do the rest as the user with the same ID as the user who
 # builds this container
 USER frontend
 WORKDIR /obs/src/api
 
-# Refresh our bundle
-RUN export NOKOGIRI_USE_SYSTEM_LIBRARIES=1; bundle install --jobs=3 --retry=3 || bundle install --jobs=3 --retry=3
-
-# Run our command
-CMD ["foreman", "start", "-f", "Procfile"]
+# Install/Refresh our gems and start OBS
+CMD bundle install --clean && foreman start -f Procfile
```

Comment:
But this will install the bundle every time I boot the app? Why? What's wrong with `docker-compose up --build` if you notice that your bundle is outdated?

Reply:
@hennevogel it only installs gems that are not there. Or, in case the Gemfile changed it installs the new version and removes the old.
That's also why I add this here. Right now, when the Gemfile changes, this container won't start anymore, because the rails app foreman starts needs the new gems and crashes.

> What's wrong with docker-compose up --build if you notice that your bundle is outdated?

I don't want to rebuild a new container every time the Gemfile changes, be it that I changed it or I just switched the branch.
To me this (the gems obs needs) is similar to the data we store in the db or the changing state of our code that we mount within the container. We want it to change frequently without having to rebuild the container each time, at least I would like to.


Thinking about it the ideal, or at least better, solution might be to stop using foreman within containers and split the different services we start there into different containers. But that would be much more effort.

Reply:
> it only installs gems that are not there

This will potentially be a lot (depending on the age of your frontend-base container) and it will make this network dependent.

> I don't want to rebuild a new container every time the Gemfile changes, be it that I changed it or I just switched the branch.

I understand, but why?

> Thinking about it the ideal, or at least better, solution might be to stop using foreman within containers and split the different services we start there into different containers. 

You can do this, of course. But controlling multiple processes is the whole point of foreman :-) So again, why would you do that?

Reply:
> hennevogel 23 minutes ago  Owner
> > it only installs gems that are not there
>
> This will potentially be a lot (depending on the age of your frontend-base container) and it will make this network dependent.

Building a new container right now also requires bundling gem and thus is network dependent. And in that scenario we truly install all gems (aka "a lot").

>> I don't want to rebuild a new container every time the Gemfile changes, be it that I changed it or I just switched the branch.

> I understand, but why?

Because it takes to much time to rebuild the whole container each time the Gemfile get's touched. And I don't need to.
Apparently there is a strong dependency between our code and the gems we use.
Right now we treat our gems as part of our container, while the code is treated like an external source that we just mount inside our container.
This doesn't seem to reflect the dependency we have. I would like to change it. Or at least allow to opt out of it.

>> Thinking about it the ideal, or at least better, solution might be to stop using foreman within containers and split the different services we start there into different containers.

> You can do this, of course. But controlling multiple processes is the whole point of foreman :-) So again, why would you do that?

And the whole point of docker / containerizing is to run multiple processes / services in separate containers. :-)

Reply:
> Building a new container right now also requires bundling gem and thus is network dependent.

Jupp, but once it's build I can do whatever I want without network. 

> And in that scenario we truly install all gems (aka "a lot").

Nope, only the ones not cached in the frontend-base container.

Reply:
> Because it takes to much time to rebuild the whole container each time the Gemfile get's touched. 

Huh? The same amount of time it takes to fetch the gems (as the rest of the intermediate containers are cached by docker). Both will just execute the same thing.

> Right now we treat our gems as part of our container, while the code is treated like an external source that we just mount inside our container.

That's how it's set up yes, and for good reasons. Gems and dependencies, especially C extensions, are different for every base OS....

Reply:
>> You can do this, of course. But controlling multiple processes is the whole point of foreman :-) So again, why would you do that?

> And the whole point of docker / containerizing is to run multiple processes / services in separate containers. :-)

Well it's certainly okay to have more processes in the container as long as you have one controlling every other process. That is the point, not the amount of processes. 

And of course you could also build many `frontend` like containers now, all having to deal with the bundle of dependencies that need to be refreshed all the time, but why would you do that?

There is no speed gain. There is not gain in simplicity. There is no need to split processes. There is a new  dependency on the network for running this...